### ISSUE TITLE
Use setBackgroundColor() on Src and Dst Programs from a VT Correlator

### ISSUE THREAD

**Issue by 0x6d696368** [NationalSecurityAgency/ghidra#1150]
Is it possible to color program listings in the Source and Destination Programs, for example via `colorService.setBackgroundColor(myAddressView, new Color(200,160,160));`, from within a Version Tracking Correlator.
I tried to get a `ColorizingService` in the `ProgramCorrelator` constructor via `this.colorService = serviceProvider.getService(ColorizingService.class);` but that returns `null`.

I can place comments and other changes in the Source and Destination Programs, though.

Colorization does not work, however, because it doesn't seem to be handled by the `Program` but rather the tool. Any idea how I can access the Source and Destination Tools from a Correlator and use `setBackgroundColor()`?

As a workaround I will write a script instead. But it would be nice to be able to color detected function changes from within a VT Session and thus get a ghetto version of the BinDiff graph diff view by coloring changes and then using Ghidra's normal graph view.

**Comment by 0x6d696368** [comment]
I forgot to ask the most important question: **Are Correlators even supposed to change programs by coloring them?** My guess would be no, and the proper place to add a graph diff feature in Ghidra would be the existing Differ View.

**Comment by dragonmacher** [comment]
**TL;DR**
1) Correlators currently do not modifify the programs
2) Comparisons (usually via highlighting) are currently done when the user selects matches in the main window
3) Version Tracking is relatively immature 
4) Work like yours is what VT needs to become a useful tool


In the original vision for VT, the correlators were intended to identify matches between the 2 programs, reporting them to the main VT tool window.   The source and destination tools exist to provide context.    The user will use the main tool window to examine and accept/reject matches as desired.  This processes of accepting matches will update the destination program.    In this line of thinking, the coloring would not be done by the correlators, but by the viewers that are showing the selected match in the main window, along with any of its markup items.    Further, it is these views in the main window (not the source and destination tools) that currently display the specific differences for the selected match.

Off the top of my head, the only downside I see to correlators modifying the programs is that the user has no easy way to remove the edits.  In this case, they may wish for the colors to come and go as they click individual matches or markkup items.

I view VT as a nascent tool, with a maturity level somewhere around version 1.2.   That is, the tool is more robust than a prototype, but it has not yet become feature complete.   I see a future for VT where the tool looks and functions completely differently than it does today.  For example, the current expected workflow is not intuitive, as the current main window is a bit noisy.   A I think we still have unanswered questions about the overall design, such as yours, that when answered will help morph tool into a mature, usable product.  

Ultimately, for VT to mature, we need to establish 1) that the tool has value for our users and then 2) we need a collection of Use Cases/Users Stories to help develop the missing features and workflow.

**Comment by 0x6d696368** [comment]
Thank you very much for Ghidra and taking your time!

> In this line of thinking, the coloring would not be done by the correlators, but by the viewers that are showing the selected match in the main window, along with any of its markup items. Further, it is these views in the main window (not the source and destination tools) that currently display the specific differences for the selected match.

That's what I figured is the technical correct place this should be implemented. I opened a feature request for that #1154 

> I view VT as a nascent tool, with a maturity level somewhere around version 1.2. That is, the tool is more robust than a prototype, but it has not yet become feature complete.

I find VT pretty useful. The lack of a correlator that shows how different two functions are was a bit confusing at first, but I implemented a prototype for such a correlator myself: https://github.com/threatrack/ghidra-patchdiff-correlator

The only thing for my purpose (patch diffing) that was missing was a graph based diffing view, as that makes it very easy to make sense of the changes.

> I see a future for VT where the tool looks and functions completely differently than it does today.
> For example, the current expected workflow is not intuitive, as the current main window is a bit noisy.

To me the workflow was not intuitve, because I didn't get it. But it basically uses the same principle of BinDiff to first find fixpoints between binaries (Exact Program correlators) then expand them by looking for callgraph or other relationships (Reference correlators).

I must say once I got the concept, I quite like it. It's a very smart design.

> A I think we still have unanswered questions about the overall design, such as yours, that when answered will help morph tool into a mature, usable product.

Well to be honest, this was actually a question relating to how to most quickly (and inappropriately) implement a clutch to get a Graph View diff.

> Ultimately, for VT to mature, we need to establish 1) that the tool has value for our users and then 2) we need a collection of Use Cases/Users Stories to help develop the missing features and workflow.

On 1) It has value to me. It diffs Windows updates nicely.
On 2) To me the only thing immediately missing was/is Graph View diffs.

I will add separate feature requests for missing features in VT.
